File: services/backend/pkg/gamification/grade.go
// Grade tiers — D → S. Pure function from total XP → letter+suffix.
// Phase 8 / M3.
package gamification

// GradeTiers — strictly ascending по min_xp. Сравнение через linear scan.
// 9 ступеней (D, D+, C, C+, B, B+, A, A+, S) — соответствует UI design.
type gradeTier struct {
	minXP int
	label string
}

var tiers = []gradeTier{
	{0, "D"},
	{100, "D+"},
	{250, "C"},
	{500, "C+"},
	{1000, "B"},
	{2000, "B+"},
	{3500, "A"},
	{5500, "A+"},
	{10000, "S"},
}

// GradeForXP — возвращает label для текущего total XP.
func GradeForXP(totalXP int) string {
	if totalXP < 0 {
		return "D"
	}
	out := "D"
	for _, t := range tiers {
		if totalXP >= t.minXP {
			out = t.label
		} else {
			break
		}
	}
	return out
}

// NextGrade — следующая ступень + сколько XP до неё. Возвращает ("", 0)
// если уже на максимуме (S).
func NextGrade(totalXP int) (label string, xpRemaining int) {
	for _, t := range tiers {
		if totalXP < t.minXP {
			return t.label, t.minXP - totalXP
		}
	}
	return "", 0
}

File: apps/mobile-rn/src/modules/gamification/domain/grade.ts
// Grade tiers — D → S. TS mirror Go pkg/gamification/grade.go.
// Phase 8 / M3.

export type Grade =
  | 'D' | 'D+'
  | 'C' | 'C+'
  | 'B' | 'B+'
  | 'A' | 'A+'
  | 'S';

type Tier = { minXP: number; label: Grade };

const TIERS: ReadonlyArray<Tier> = [
  { minXP: 0,     label: 'D'  },
  { minXP: 100,   label: 'D+' },
  { minXP: 250,   label: 'C'  },
  { minXP: 500,   label: 'C+' },
  { minXP: 1000,  label: 'B'  },
  { minXP: 2000,  label: 'B+' },
  { minXP: 3500,  label: 'A'  },
  { minXP: 5500,  label: 'A+' },
  { minXP: 10000, label: 'S'  },
];

export function gradeForXP(totalXP: number): Grade {
  if (!Number.isFinite(totalXP) || totalXP < 0) return 'D';
  let out: Grade = 'D';
  for (const t of TIERS) {
    if (totalXP >= t.minXP) {
      out = t.label;
    } else {
      break;
    }
  }
  return out;
}

/**
 * Next grade tier + XP до неё. Возвращает null если уже на S (максимум).
 */
export function nextGrade(totalXP: number): { label: Grade; xpRemaining: number } | null {
  for (const t of TIERS) {
    if (totalXP < t.minXP) {
      return { label: t.label, xpRemaining: t.minXP - totalXP };
    }
  }
  return null;
}
